File: server/src/repositories/financeRepository.js
const db = require('../db');

class FinanceRepository {
  // Accounts
  getAccounts(userId) {
    return db.prepare('SELECT * FROM accounts WHERE user_id = ? ORDER BY is_default DESC, id ASC').all(userId);
  }

  insertAccount(userId, name, bank, numberMasked, type, balance, isDefault, now) {
    return db.prepare(`
      INSERT INTO accounts (user_id, account_name, bank_name, account_number_masked, account_type, opening_balance, current_balance, is_default, created_at, updated_at)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `).run(userId, name, bank, numberMasked, type, balance, balance, isDefault ? 1 : 0, now, now);
  }

  getAccountById(accountId, userId) {
    return db.prepare('SELECT * FROM accounts WHERE id = ? AND user_id = ?').get(accountId, userId);
  }

  // Categories
  getCategories(userId) {
    return db.prepare('SELECT * FROM categories WHERE user_id = ? ORDER BY type ASC, name ASC').all(userId);
  }

  insertCategory(userId, name, type, icon, color, now) {
    return db.prepare(`
      INSERT INTO categories (user_id, name, type, icon, color, is_default, created_at)
      VALUES (?, ?, ?, ?, ?, 0, ?)
    `).run(userId, name, type, icon, color, now);
  }

  getCategoryById(catId, userId) {
    return db.prepare('SELECT * FROM categories WHERE id = ? AND user_id = ?').get(catId, userId);
  }

  // Rules
  getRules(userId) {
    return db.prepare('SELECT * FROM categorization_rules WHERE user_id = ? ORDER BY priority DESC').all(userId);
  }

  insertRule(userId, pattern, categoryId, categoryName, transactionType, priority, matchType, now) {
    return db.prepare(`
      INSERT INTO categorization_rules (user_id, pattern, category_id, category_name, transaction_type, priority, match_type, is_active, created_at)
      VALUES (?, ?, ?, ?, ?, ?, ?, 1, ?)
    `).run(userId, pattern, categoryId, categoryName, transactionType, priority, matchType, now);
  }

  getRuleById(ruleId, userId) {
    return db.prepare('SELECT * FROM categorization_rules WHERE id = ? AND user_id = ?').get(ruleId, userId);
  }

  // Transactions
  getTransactions(userId) {
    return db.prepare('SELECT * FROM transactions WHERE user_id = ? ORDER BY transaction_date DESC, id DESC').all(userId);
  }

  getRecentTransactions(userId, limit) {
    return db.prepare('SELECT * FROM transactions WHERE user_id = ? ORDER BY transaction_date DESC LIMIT ?').all(userId, limit);
  }

  getTransactionById(txId, userId) {
    return db.prepare('SELECT * FROM transactions WHERE id = ? AND user_id = ?').get(txId, userId);
  }

  insertTransaction(userId, accountId, date, description, debit, credit, amount, type, categoryId, categoryName, notes, source, now) {
    return db.prepare(`
      INSERT INTO transactions (user_id, account_id, transaction_date, description, debit_amount, credit_amount, amount, transaction_type, category_id, category_name, notes, source, created_at, updated_at)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    `).run(userId, accountId, date, description, debit, credit, amount, type, categoryId, categoryName, notes, source, now, now);
  }

  // Loans
  getLoans(userId) {
    return db.prepare('SELECT * FROM loans WHERE user_id = ? ORDER BY id DESC').all(userId);
  }

  getActiveLoans(userId) {
    return db.prepare("SELECT * FROM loans WHERE user_id = ? AND status != 'PAID'").all(userId);
  }

  getLoanById(loanId, userId) {
    return db.prepare('SELECT * FROM loans WHERE id = ? AND user_id = ?').get(loanId, userId);
  }

  insertLoan(userId, borrowerLenderName, loanType, amount, interestRate, dueDate, notes, now) {
    return db.prepare(`
      INSERT INTO loans (user_id, borrower_lender_name, loan_type, amount, total_repaid, remaining_amount, interest_rate, due_date, notes, status, created_at, updated_at)
      VALUES (?, ?, ?, ?, 0.0, ?, ?, ?, ?, 'ACTIVE', ?, ?)
    `).run(userId, borrowerLenderName, loanType, amount, amount, interestRate, dueDate, notes, now, now);
  }

  updateLoanStatus(loanId, userId, totalRepaid, remaining, status, now) {
    return db.prepare('UPDATE loans SET total_repaid = ?, remaining_amount = ?, status = ?, updated_at = ? WHERE id = ? AND user_id = ?')
      .run(totalRepaid, remaining, status, now, loanId, userId);
  }

  // Repayments
  insertRepayment(userId, loanId, amount, date, paymentMethod, notes, now) {
    return db.prepare(`
      INSERT INTO loan_repayments (user_id, loan_id, amount, repayment_date, payment_method, notes, created_at)
      VALUES (?, ?, ?, ?, ?, ?, ?)
    `).run(userId, loanId, amount, date, paymentMethod, notes, now);
  }

  getRepaymentsSum(loanId, userId) {
    return db.prepare('SELECT SUM(amount) as total FROM loan_repayments WHERE loan_id = ? AND user_id = ?').get(loanId, userId);
  }

  // Company Expenses
  getCompanyExpenses(userId) {
    return db.prepare('SELECT * FROM company_expenses WHERE user_id = ? ORDER BY expense_date DESC').all(userId);
  }

  insertCompanyExpense(userId, date, title, amount, categoryName, notes, now) {
    return db.prepare(`
      INSERT INTO company_expenses (user_id, expense_date, title, amount, category_name, notes, is_reimbursed, created_at)
      VALUES (?, ?, ?, ?, ?, ?, 0, ?)
    `).run(userId, date, title, amount, categoryName, notes, now);
  }
}

module.exports = new FinanceRepository();

File: app/src/main/java/com/example/repository/FinanceRepository.kt
package com.example.repository

import com.example.data.local.AppDatabase
import com.example.data.local.entity.*
import com.example.features.rules.CategorizationEngine
import com.example.features.rules.CategorizationResult
import kotlinx.coroutines.flow.Flow

class FinanceRepository(private val database: AppDatabase) {
    private val accountDao = database.accountDao()
    private val categoryDao = database.categoryDao()
    private val ruleDao = database.categorizationRuleDao()
    private val transactionDao = database.transactionDao()
    private val loanDao = database.loanDao()
    private val loanRepaymentDao = database.loanRepaymentDao()
    private val companyExpenseDao = database.companyExpenseDao()
    private val profileDao = database.userProfileDao()
    private val undoDao = database.undoDao()

    // Undo History
    val undoHistory: Flow<List<UndoHistoryEntity>> = undoDao.getRecentUndoActions()
    suspend fun insertUndoAction(action: UndoHistoryEntity) = undoDao.insertUndoAction(action)
    suspend fun getLatestUndoAction() = undoDao.getLatestUndoAction()
    suspend fun getUndoActionById(id: String) = undoDao.getUndoActionById(id)
    suspend fun deleteUndoAction(id: String) = undoDao.deleteUndoAction(id)
    suspend fun clearUndoHistory() = undoDao.clearUndoHistory()

    // Profile & Settings
    val userProfile: Flow<UserProfileEntity?> = profileDao.getUserProfile()
    suspend fun updateProfile(profile: UserProfileEntity) = profileDao.insertOrUpdateProfile(profile)
    suspend fun updateThemePreference(isDark: Boolean) = profileDao.updateThemePreference(isDark)
    suspend fun updateDashboardCardsConfig(config: String) = profileDao.updateDashboardCardsConfig(config)

    // Accounts
    val allAccounts: Flow<List<AccountEntity>> = accountDao.getAllAccounts()
    suspend fun getAccountById(id: Long) = accountDao.getAccountById(id)
    suspend fun insertAccount(account: AccountEntity) = accountDao.insertAccount(account)
    suspend fun updateAccount(account: AccountEntity) = accountDao.updateAccount(account)
    suspend fun deleteAccount(id: Long) = accountDao.deleteAccount(id)

    // Categories
    val allCategories: Flow<List<CategoryEntity>> = categoryDao.getAllCategories()
    suspend fun getCategoriesByType(type: CategoryType) = categoryDao.getCategoriesByType(type)
    suspend fun insertCategory(category: CategoryEntity) = categoryDao.insertCategory(category)
    suspend fun updateCategory(category: CategoryEntity) = categoryDao.updateCategory(category)
    suspend fun deleteCategory(id: Long) = categoryDao.deleteCategory(id)

    // Categorization Rules
    val allRules: Flow<List<CategorizationRuleEntity>> = ruleDao.getAllRules()
    suspend fun getActiveRulesList() = ruleDao.getActiveRulesList()
    suspend fun insertRule(rule: CategorizationRuleEntity) = ruleDao.insertRule(rule)
    suspend fun updateRule(rule: CategorizationRuleEntity) = ruleDao.updateRule(rule)
    suspend fun toggleRule(id: Long, isActive: Boolean) = ruleDao.toggleRuleActive(id, isActive)
    suspend fun deleteRule(id: Long) = ruleDao.deleteRule(id)

    // Transactions
    val allTransactions: Flow<List<TransactionEntity>> = transactionDao.getAllTransactions()
    fun getRecentTransactions(limit: Int = 10): Flow<List<TransactionEntity>> = transactionDao.getRecentTransactions(limit)
    suspend fun getTransactionById(id: Long) = transactionDao.getTransactionById(id)

    fun searchTransactions(
        query: String,
        type: TransactionType?,
        categoryName: String?,
        accountId: Long?,
        startDate: String?,
        endDate: String?
    ) = transactionDao.searchTransactions(query, type, categoryName, accountId, startDate, endDate)

    suspend fun insertTransaction(transaction: TransactionEntity): Long {
        val id = transactionDao.insertTransaction(transaction)
        recalculateAccountBalance(transaction.accountId)
        return id
    }

    suspend fun insertTransactions(transactions: List<TransactionEntity>): List<Long> {
        val ids = transactionDao.insertTransactions(transactions)
        val accountIds = transactions.map { it.accountId }.distinct()
        accountIds.forEach { recalculateAccountBalance(it) }
        return ids
    }

    suspend fun updateTransaction(transaction: TransactionEntity) {
        transactionDao.updateTransaction(transaction)
        recalculateAccountBalance(transaction.accountId)
    }

    suspend fun updateTransactionCategory(id: Long, categoryId: Long?, categoryName: String, transactionType: TransactionType) {
        transactionDao.updateTransactionCategory(id, categoryId, categoryName, transactionType)
    }

    suspend fun deleteTransaction(id: Long) {
        val tx = transactionDao.getTransactionById(id)
        transactionDao.deleteTransaction(id)
        tx?.let { recalculateAccountBalance(it.accountId) }
    }

    // Loans & Repayments
    val allLoans: Flow<List<LoanEntity>> = loanDao.getAllLoans()
    val activeLoans: Flow<List<LoanEntity>> = loanDao.getActiveLoans()
    suspend fun getLoanById(id: Long) = loanDao.getLoanById(id)
    suspend fun insertLoan(loan: LoanEntity): Long = loanDao.insertLoan(loan)
    suspend fun updateLoan(loan: LoanEntity) = loanDao.updateLoan(loan)
    suspend fun deleteLoan(id: Long) = loanDao.deleteLoan(id)

    fun getRepaymentsForLoan(loanId: Long): Flow<List<LoanRepaymentEntity>> = loanRepaymentDao.getRepaymentsForLoan(loanId)

    suspend fun recordLoanRepayment(loanId: Long, amount: Double, repaymentDate: String, method: String, notes: String): Long {
        val repayment = LoanRepaymentEntity(
            loanId = loanId,
            amount = amount,
            repaymentDate = repaymentDate,
            paymentMethod = method,
            notes = notes
        )
        val repaymentId = loanRepaymentDao.insertRepayment(repayment)

        // Update Loan progress
        val loan = loanDao.getLoanById(loanId)
        if (loan != null) {
            val totalRepaid = (loanRepaymentDao.getTotalRepaidForLoan(loanId) ?: 0.0)
            val remaining = (loan.amount - totalRepaid).coerceAtLeast(0.0)
            val status = when {
                remaining <= 0.0 -> LoanStatus.PAID
                totalRepaid > 0.0 -> LoanStatus.PARTIALLY_PAID
                else -> LoanStatus.ACTIVE
            }
            loanDao.updateLoanRepaymentProgress(loanId, totalRepaid, remaining, status)
        }
        return repaymentId
    }

    suspend fun deleteRepayment(repaymentId: Long, loanId: Long) {
        loanRepaymentDao.deleteRepayment(repaymentId)
        val loan = loanDao.getLoanById(loanId)
        if (loan != null) {
            val totalRepaid = (loanRepaymentDao.getTotalRepaidForLoan(loanId) ?: 0.0)
            val remaining = (loan.amount - totalRepaid).coerceAtLeast(0.0)
            val status = when {
                remaining <= 0.0 -> LoanStatus.PAID
                totalRepaid > 0.0 -> LoanStatus.PARTIALLY_PAID
                else -> LoanStatus.ACTIVE
            }
            loanDao.updateLoanRepaymentProgress(loanId, totalRepaid, remaining, status)
        }
    }

    // Company Expenses
    val allCompanyExpenses: Flow<List<CompanyExpenseEntity>> = companyExpenseDao.getAllCompanyExpenses()
    suspend fun insertCompanyExpense(expense: CompanyExpenseEntity) = companyExpenseDao.insertCompanyExpense(expense)
    suspend fun updateCompanyExpense(expense: CompanyExpenseEntity) = companyExpenseDao.updateCompanyExpense(expense)
    suspend fun updateReimbursementStatus(id: Long, isReimbursed: Boolean) = companyExpenseDao.updateReimbursementStatus(id, isReimbursed)
    suspend fun deleteCompanyExpense(id: Long) = companyExpenseDao.deleteCompanyExpense(id)

    // Automatic Categorization Helper
    suspend fun autoCategorize(description: String, isCredit: Boolean = false): CategorizationResult {
        val activeRules = ruleDao.getActiveRulesList()
        return CategorizationEngine.categorize(description, activeRules, isCredit)
    }

    // Balance recalculation helper
    private suspend fun recalculateAccountBalance(accountId: Long) {
        val account = accountDao.getAccountById(accountId) ?: return
        // Keep starting balance + sum of income/transfers in - sum of expenses/transfers out
        // Or if user added transactions with balances, update latest
    }
}
